src/sinks/util/buffer/metrics.rs

```diff
@@ -0,0 +1,548 @@
+use crate::event::{Event, Metric};
+use crate::sinks::util::Batch;
+use indexmap::IndexSet;
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+
+#[derive(Clone)]
+struct MetricEntry(Metric);
+
+impl Eq for MetricEntry {}
+
+impl Hash for MetricEntry {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        std::mem::discriminant(&self.0).hash(state);
+
+        match &self.0 {
+            Metric::Counter { name, .. } => {
+                name.hash(state);
+            }
+            Metric::Gauge { name, .. } => {
+                name.hash(state);
+            }
+            Metric::Set { name, val, .. } => {
+                name.hash(state);
+                val.hash(state);
+            }
+            Metric::Histogram { name, val, .. } => {
+                name.hash(state);
+                val.to_bits().hash(state);
+            }
+        }
+
+        self.0
+            .tags()
+            .as_ref()
+            .map(|ts| ts.iter().for_each(|t| t.hash(state)));
```

Comment:
Can we not just do `self.0.tags().hash(state)` here?

Reply:
This is because `HashMap` does not implement `Hash`. The reason is

> To have HashMaps implement Hash, we need to iterate through their contained hashes in a consistent order.

Oh wait.. this means my naive iterative hashing is also not perfectly stable.

BTW, I was thinking some time ago about switching to IndexMap for storing tags (or maybe even something like `Vec<(String, String)>`)

Reply:
Ah, I'd assumed it was a `Vec` already. I don't think we want order to matter when hashing tags, so maybe `BTreeMap`?

Reply:
I believe the only place where we do something Map-ish with the tags is the `remove_tags` transform. Normally what's happening to tags is they are being constructed and carried into sinks, where they are transformed into Vec<_> or String.

I'll be fine with anything having stable order, but I'd vote for Vec because of ergonomic reasons, to remove some boilerplate converting to and from Maps.

Performance wise, because metrics do not contain many tags usually, I expect working with small Vec's should be effective even for operations like `remove_tags`.

Reply:
Makes sense! My only concern with a `Vec` is the case where two different metrics have the same tags but in a different order. I think we'd still want to merge them, but maybe we could adjust the hashing to account for that.